b-harvest/minievm
# feat: filter apis

## Proposed changes

This PR implements JSON-RPC APIs including `eth_newFilter`, `eth_newBlockFilter`, `eth_getFilterLogs`, `eth_getFilterChanges`, and `eth_getLogs`.

Fixed a bug where `extractLogsFromEvents` always returns an empty array.


## Review Points

### 1. the information held by the indexer in the Logs

Verify whether the data missing in the Logs is correctly retrieved from the indexed blockHeader and txs.

### 2. Functionality of filter APIs

Check if the implemented JSON-RPC APIs are working correctly.

## How to test this PR

1. Clone the repository and switch to the relevant branch.
2. Build the project using `make install`.
3. Setup a genesis for test
4. Start the Minitia EVM node using `minitiad start`
5. Make a transaction to check events
6. Use a JSON-RPC client to call the implemented methods and verify their responses.

### Example Test Commands

- `eth_newFilter` 

```bash
$  curl -X POST  -H "Content-Type: application/json" --data '{"jsonrpc":"2.0","method":"eth_newFilter","params":[{"fromBlock": "0x2390", "toBlock": "0x2399", "address": "0x42851b9999ac002394945A39B301c0d474c97417","topics": ["0xddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef"]}],"id":1}'  http://localhost:8545 

{"jsonrpc":"2.0","id":1,"result":"0x129679eccf963fd15e01f6504f6d060a"}
```

- `eth_getFilterLogs` 

```bash
$  curl -X POST  -H "Content-Type: application/json" --data '{"method":"eth_getFilterLogs","params":["0x129679eccf963fd15e01f6504f6d060a"],"id":1,"jsonrpc":"2.0"}' http://localhost:8545 

{"jsonrpc":"2.0","id":1,"result":[{"address":"0x42851b9999ac002394945a39b301c0d474c97417","topics":["0xddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef","0x000000000000000000000000c6fe5d33615a1c52c08018c47e8bc53646a0e101","0x00000000000000000000000025051c50bb6f302a7bd89100bead641e65da743b"],"data":"0x0000000000000000000000000000000000000000000000000000000000000457","blockNumber":"0x2398","transactionHash":"0x80dcea3db9c181748bce112c760102b2f4d6ac7c25d514ec4ed16656b8fefdb6","transactionIndex":"0x1","blockHash":"0x6ebdf8f6a2ccf6c9b9aead31589524949e8e63c16e938f07100725e02c219432","logIndex":"0x0","removed":false}]}
```

- `eth_newBlockFilter` 

```bash
$  curl -X POST  -H "Content-Type: application/json" --data '{"jsonrpc":"2.0","method":"eth_newBlockFilter","params":[],"id":1}' http://localhost:8545 

{"jsonrpc":"2.0","id":1,"result":"0x47bf8617648970968c2481aa433c2836"}
```

- `eth_getFilterChanges` 

```bash
$  curl -X POST  -H "Content-Type: application/json" --data '{"method":"eth_getFilterChanges","params":["0x47bf8617648970968c2481aa433c2836"],"id":1,"jsonrpc":"2.0"}' http://localhost:8545 

{"jsonrpc":"2.0","id":1,"result":["0x44c80c0041d93a5866a448a0d48762f8406e9df6354b099c2469137181d5649a"]}
```

- `eth_uninstallFilter` 

```bash
$  curl -X POST  -H "Content-Type: application/json" --data '{"method":"eth_uninstallFilter","params":["0x47bf8617648970968c2481aa433c2836"],"id":1,"jsonrpc":"2.0"}' http://localhost:8545 

{"jsonrpc":"2.0","id":1,"result":true}

$  curl -X POST  -H "Content-Type: application/json" --data '{"method":"eth_uninstallFilter","params":["0x47bf8617648970968c2481aa433c2836"],"id":1,"jsonrpc":"2.0"}' http://localhost:8545 

{"jsonrpc":"2.0","id":1,"error":{"code":-32000,"message":"filter 0x47bf8617648970968c2481aa433c2836 not found"}}
```

- `eth_uninstallFilter` 

```bash
$  curl -X POST  -H "Content-Type: application/json" --data '{"method":"eth_getLogs","params":[{"blockHash":"0x6ebdf8f6a2ccf6c9b9aead31589524949e8e63c16e938f07100725e02c219432"}],"id":1,"jsonrpc":"2.0"}' http://localhost:8545 

{"jsonrpc":"2.0","id":1,"result":[{"address":"0x42851b9999ac002394945a39b301c0d474c97417","topics":["0xddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef","0x000000000000000000000000c6fe5d33615a1c52c08018c47e8bc53646a0e101","0x00000000000000000000000025051c50bb6f302a7bd89100bead641e65da743b"],"data":"0x0000000000000000000000000000000000000000000000000000000000000457","blockNumber":"0x2398","transactionHash":"0x80dcea3db9c181748bce112c760102b2f4d6ac7c25d514ec4ed16656b8fefdb6","transactionIndex":"0x1","blockHash":"0x6ebdf8f6a2ccf6c9b9aead31589524949e8e63c16e938f07100725e02c219432","logIndex":"0x0","removed":false}]}

```

## Types of changes



## Related issues

N/A
